crates/matrix-sdk-base/src/rooms/calls.rs

```diff
@@ -0,0 +1,144 @@
+            .flat_map(|(user_id, devices)| {
+                devices.iter().map(|device| {
+                    let id = CallMemberIdentifier {
+                        user_id: user_id.clone(),
+                        device_id: device.device_id.clone(),
+                    };
+
+                    (id, device.info.clone())
+                })
+            })
+            .filter(|(_, data)| !data.expired())
+            .collect();
+
+        (!members.is_empty()).then_some(RoomCall { members })
+    }
+}
+
+/// Represents a room call (MatrixRTC session). A call consists of a set of
+/// members who participate in a call. A call is active as long as there is at
+/// least one member (participant) in it.
+#[derive(Debug)]
+pub struct RoomCall {
+    /// The members (participants) of the call as of the moment of the last
+    /// sync.
+    pub(super) members: HashMap<CallMemberIdentifier, CallMemberInfo>,
+}
+
+impl RoomCall {
+    /// Returns the time when the call was started, i.e. when the first known
+    /// member joined.
+    pub fn start_time(&self) -> MilliSecondsSinceUnixEpoch {
+        self.members
+            .values()
+            .map(|member| member.member_since)
+            .min()
+            .expect("Bug: a room call must not be instantiated with 0 members")
+    }
+
+    /// Non-expired members of a call.
+    pub fn active_members(&self) -> impl Iterator<Item = &CallMemberIdentifier> {
+        self.members.iter().filter(|(_, member)| !member.expired()).map(|(id, _)| id)
+    }
+}
+
+/// Info about a call member (participant).
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct CallMemberInfo {
+    /// The time when the member joined the call.
+    pub member_since: MilliSecondsSinceUnixEpoch,
+    /// The duration for which the member is valid starting from
+    /// [`Self::member_since`].
+    pub valid_for: Duration,
+}
+
+impl CallMemberInfo {
+    /// Returns `true` if the member is still valid, i.e. if the member joined
+    /// and its membership has not expired yet. Returns `false` otherwise.
+    pub fn expired(&self) -> bool {
+        let expires_at = self.member_since.to_system_time().map(|start| start + self.valid_for);
+        let now = MilliSecondsSinceUnixEpoch::now().to_system_time();
+        expires_at.zip(now).map(|(expires_at, now)| expires_at < now).unwrap_or(true)
+    }
+}
+
+/// A unique identifier of a call member (participant).
+/// Each call member is identified by their user ID and device ID.
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct CallMemberIdentifier {
+    /// The user ID of the call member.
+    pub user_id: OwnedUserId,
+    /// The device ID of the call member.
+    pub device_id: OwnedDeviceId,
+}
+
+/// Represents a single member of a call.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct CallMember {
+    /// The member's device ID.
+    device_id: OwnedDeviceId,
+    /// Member info.
+    info: CallMemberInfo,
+}
```

Comment:
since the device_id and the user_id are similar and both build the identifier I am wondering if we would like `CallMember` to contain neither of the two. And instead always store `CallMemberInfo` in a Map with `CallMemberIdentifier`
```
HashMap<CallMemberIdentifier, CallMemberInfo>
```
(similar to the comment above. But seeing that we have device_id in the CallMember but not user_id I think we could remove the `CallMember` struct entirely.

Reply:
Right, I tried to answer it here: https://github.com/matrix-org/matrix-rust-sdk/pull/2843#discussion_r1408143706

But this part may certainly be changed in the upcoming PRs.